espresso/environment/espresso_caff_node_test.go

```diff
@@ -0,0 +1,266 @@
+package environment_test
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"log/slog"
+	"math/big"
+	"net"
+	"net/url"
+	"testing"
+	"time"
+
+	env "github.com/ethereum-optimism/optimism/espresso/environment"
+	"github.com/ethereum-optimism/optimism/op-e2e/e2eutils/geth"
+	"github.com/ethereum-optimism/optimism/op-e2e/e2eutils/opnode"
+	"github.com/ethereum-optimism/optimism/op-e2e/e2eutils/wait"
+	"github.com/ethereum-optimism/optimism/op-e2e/system/e2esys"
+	"github.com/ethereum-optimism/optimism/op-e2e/system/helpers"
+	"github.com/ethereum-optimism/optimism/op-node/chaincfg"
+	"github.com/ethereum-optimism/optimism/op-node/rollup"
+	"github.com/ethereum-optimism/optimism/op-service/testlog"
+	"github.com/ethereum/go-ethereum/accounts/abi/bind"
+)
+
+const (
+	RoleCaffNode = "caff-node"
+)
+
+// ErrorFailedToParseSequencerPort is returned when the sequencer port
+// cannot be parsed from the espresso dev node.
+type ErrorFailedToParseSequencerPort struct {
+	Have string
+}
+
+// Error implements error
+func (e ErrorFailedToParseSequencerPort) Error() string {
+	return fmt.Sprintf("failed to parse sequencer port URL: \"%s\"", e.Have)
+}
+
+// ErrorFailedToStartCaffNodeGeth is returned when the caff node geth
+// instance fails to start.
+type ErrorFailedToStartCaffNodeGeth struct {
+	Cause error
+}
+
+// Error implements error
+func (e ErrorFailedToStartCaffNodeGeth) Error() string {
+	return fmt.Sprintf("failed to start caff node geth instance: %v", e.Cause)
+}
+
+// Unwrap allows for the root cause of the error to be extracted.
+func (e ErrorFailedToStartCaffNodeGeth) Unwrap() error {
+	return e.Cause
+}
+
+// ErrorFailedToStartCaffNodeOpNode is returned when the caff node op
+// node instance fails to start.
+type ErrorFailedToStartCaffNodeOpNode struct {
+	Cause error
+}
+
+// Error implements error
+func (e ErrorFailedToStartCaffNodeOpNode) Error() string {
+	return fmt.Sprintf("failed to start caff node op node instance: %v", e.Cause)
+}
+
+// Unwrap allows for the root cause of the error to be extracted.
+func (e ErrorFailedToStartCaffNodeOpNode) Unwrap() error {
+	return e.Cause
+}
+
+// CaffNodeInstance is a wrapper around the caff node geth instance and op node
+// instance, for the Caff Node. It is used to interact with the caff node.
+type CaffNodeInstance struct {
+	OpNode *opnode.Opnode
+	Geth   *geth.GethInstance
+}
+
+// Close closes the caff node geth instance and op node instance.
+func (c *CaffNodeInstance) Close(ctx context.Context) error {
+	return errors.Join(c.OpNode.Stop(ctx), c.Geth.Close())
+}
+
+// LaunchDecaffNode launches a caff node in the given system. It will
+// configure the caff node to connect to the given espresso dev node.
+func LaunchDecaffNode(t *testing.T, system *e2esys.System, espressoDevNode env.EspressoDevNode) (*CaffNodeInstance, error) {
```

Comment:
At some point this function should be moved to a helper file right?

Reply:
Sure, not a bad idea... I had originally tried to put it in it's own file, but was getting errors because I was referencing the `environment_test` package, which is only available in a test file.  But I realized that the other environment helpers I did are just in the `environment` package itself, so I should just continue form.

Reply:
Rebased and split the helper into its own file in the `environment` package.  Pushed in27b7362ebad9585436154c779ffb3d44e24837e6.